### ISSUE TITLE
Question about using `threading.local` with eventlet and gunicorn

### ISSUE THREAD

**Issue by bunnymatic** [eventlet/eventlet#723]
We're running a Django project with `gunicorn` and `eventlet`.  

I'd like to use `threading.local` to stash some http request data for use later in that thread.  I'm wondering if this is safe with `eventlet`.

From the docs
>Eventlet is thread-safe and can be used in conjunction with normal Python threads. The way this works is that coroutines are confined to their ‘parent’ Python thread. It’s like each thread contains its own little world of coroutines that can switch between themselves but not between coroutines in other threads.
which sounds like it might be.  

But I understand that `eventlet` will, based on reading the docs on 'How the Hubs Work', may suspend a co-routine to process another one.  Is it possible, with `gunicorn` the an http request processing may get suspended and another http request would get picked up and processed by a co-routine in that same initial thread?  And if so, then I think that means a `threading.local` in Django middleware might get shared between two requests.   Does that sound right?

Can I get away with using `threading.local` and be certain that each incoming request will get it's own thread.local space?  

If not I wonder if the co-routine processing is such that I could store a stack in the thread local so as more requests were processed by that thread I could push state onto this stack and pop it off as requests finish up.

Any pointers would be appreciated here.

Thanks

**Comment by temoto** [comment]
TL;DR: after `eventlet.monkey_patch()` (which is done by running `gunicorn -k eventlet`) just `threading.local()` will work as expected.

General idea is: green version of threading module treats each coroutine as separate thread.

Test:

```python
# app.py
import eventlet

TRY_GREEN_FLAG = True

if TRY_GREEN_FLAG:
    from eventlet.green import threading
else:
    threading = eventlet.patcher.original("threading")

local = threading.local()

def app(environ, start_response):
    unique = environ["REMOTE_PORT"]
    print("remote={} thread_id={}".format(unique, threading.get_ident()))
    start_response("200 OK", [])
    local.data = unique
    # other thread will execute here and could have a chance to overwrite local.data
    eventlet.sleep(1)
    return [repr(local.data).encode()]
```

```shell
mkdir tmp
cd tmp
python3 -m venv .
bin/pip install -U pip wheel
bin/pip install eventlet
# patch https://github.com/benoitc/gunicorn/pull/2581 is merged but not yet released on pypi
bin/pip install 'https://github.com/benoitc/gunicorn/archive/refs/heads/master.zip'
bin/gunicorn -k eventlet app:app

# TRY_GREEN_FLAG=False
( u='http://127.0.0.1:8000/' ; curl "$u" & sleep 0.1s ; curl "$u" & sleep 0.1s )
'60177''60177'
# native threading.local() is same for all requests

# TRY_GREEN_FLAG=True
( u='http://127.0.0.1:8000/' ; curl "$u" & sleep 0.1s ; curl "$u" & sleep 0.1s )
'60178''60179'
# green threading.local() is unique for each coroutine
```

Have a great day.

**Comment by bunnymatic** [comment]
Indeed I will.  Thank you so much for the quick response and proof.

**Comment by dimas-c5** [comment]
Super helpful, thank you!!
